docs/proposals/cluster-deployment.md

```diff
@@ -0,0 +1,204 @@
+4. Create firewall rule to allow ```https``` to master
+5. Create persistent disk for master
+6. Create static IP address for master
+7. Create master machine
+8. Create node machines
+9. Install docker on all machines
+
+**Exit critera**:
+
+1. Can ```ssh``` to all machines and run a test docker image
+2. Can ```ssh``` to master and nodes and ping other machines
+
+## Step 2: Generate certificates
+
+**Objective**: Generate security certificates used to configure secure communication between client, master and nodes
+
+TODO: Enumerate ceritificates which have to be generated.
+
+## Step 3: Deploy master
+
+**Objective**: Run kubelet and all the required components (e.g. etcd, apiserver, scheduler, controllers) on the master machine.
+
+**Substeps**:
+
+1. copy certificates
+2. copy manifests for static pods:
+	1. etcd
+	2. apiserver, controller manager, scheduler
+3. run kubelet in docker container (configuration is read from apiserver Config object)
+4. run kubelet-checker in docker container
+
+**v1.2 simplifications**:
+
+1. kubelet-runner.sh - we will provide a custom docker image to run kubelet; it will contain
+kubelet binary and will run it using ```nsenter``` to workaround problem with mount propagation
+1. kubelet config file - we will read kubelet configuration file from disk instead of apiserver; it will
+be generated locally and copied to all nodes.
+
+**Exit criteria**:
+
+1. Can run basic API calls (e.g. create, list and delete pods) from the client side (e.g. replication
+controller works - user can create RC object and RC manager can create pods based on that)
+2. Critical master components works:
+  1. scheduler
+  2. controller manager
+
+## Step 4: Deploy nodes
+
+**Objective**: Start kubelet on all nodes and configure kubernetes network.
+Each node can be deployed separately and the implementation should make it ~impossible to change this assumption.
+
+### Step 4.1: Run kubelet
+
+**Substeps**:
+
+1. copy certificates
+2. run kubelet in docker container (configuration is read from apiserver Config object)
+3. run kubelet-checker in docker container
+
+**v1.2 simplifications**:
+
+1. kubelet config file - we will read kubelet configuration file from disk instead of apiserver; it will
+be generated locally and copied to all nodes.
+
+**Exit critera**:
+
+1. All nodes are registered, but not ready due to lack of kubernetes networking.
+
+### Step 4.2: Setup kubernetes networking
+
+**Objective**: Configure the Kubernetes networking to allow routing requests to pods and services.
+
+To keep default setup consistent across open source deployments we will use Flannel to configure
+kubernetes networking. However, implementation of this step will allow to easily plug in different
+network solutions.
+
+**Substeps**:
+
+1. copy manifest for flannel server to master machine
+2. create a daemonset with flannel daemon (it will read assigned CIDR and configure network appropriately).
```

Comment:
Currently flannel acquires it's own cidr and the nodecontroller is told to back off from cidr allocation when --networking=flannel. In general I'd like to support both modes, advantages of allowing third party cidr allocation include eg: being able to use non-continguous cidr ranges. no reason all that logic should live in nodecontroller, which will just do the simple 'take this /16 and carve out contiguous /24s' thing. 

Not sure how this will get implemented down the line, so best to be vague about flannel acquiring/using the allocated cidr, if possible.